index/scorch/introducer.go

```diff
@@ -124,23 +125,29 @@ func (s *Scorch) introduceSegment(next *segmentIntroduction) error {
 				return err
 			}
 		}
-		newSnapshot.segment[i] = &SegmentSnapshot{
+		newSnapshot.segment[index] = &SegmentSnapshot{
 			id:         s.root.segment[i].id,
 			segment:    s.root.segment[i].segment,
 			cachedDocs: s.root.segment[i].cachedDocs,
 		}
-		s.root.segment[i].segment.AddRef()
 
 		// apply new obsoletions
 		if s.root.segment[i].deleted == nil {
-			newSnapshot.segment[i].deleted = delta
+			newSnapshot.segment[index].deleted = delta
 		} else {
-			newSnapshot.segment[i].deleted = roaring.Or(s.root.segment[i].deleted, delta)
+			newSnapshot.segment[index].deleted = roaring.Or(s.root.segment[i].deleted, delta)
 		}
 
-		newSnapshot.offsets[i] = running
-		running += s.root.segment[i].Count()
-
+		// check for live size before copying
```

Comment:
here the comment says "before copying", but you already did "copy" on line 128 right?
it seems like the else block then tries to undo this? but the pattern here

x = append(x[:offset],x[offset+1:]...)

is used in a general purpose remove element from slice operation, but i don't understand why we use that here, we're going through the old root, and building the new snapshot slices one element at a time, there shouldn't be anything after "index" in the new segment yet right?

Reply:
yea..copying completes by setting the right ref count and setting the remaining fields. But before this final step if the live size happen to be zero, then we need to unset the assignment made at line:128 at index position.  For eg: what if the index assignment  made  to the last element/segment which later turned to have a zero live size (line:142). In that case,  we have to remove the last zero sized segment at index position. 
I can see that its due to the way this is coded up.. May be we can assign that segment to a temp variable, apply deletes , check live size and insert into the new snapshot slice.